From c373685175771263700162ae9c9bdbfa49a12620 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Julian=20R=C3=BCth?= <julian.rueth@fsfe.org>
Date: Mon, 8 Dec 2025 02:25:04 +0200
Subject: [PATCH 04/14] Revert system_dirs.diff

this patch from upstream should only be applied on Linux builds.
However, probably because the tarball is generated on Linux, it's
applied in the source tarball already.

In any case, it is unclear whether this patch is needed for Linux in
conda-forge. See
https://github.com/conda-forge/cppyy-cling-feedstock/pull/40#issuecomment-975781900
and https://github.com/wlav/cppyy/issues/156 for what seems to be the
relevant discussion.

We unapply this patch here unconditionally since it is wrong on macOS
and it also complicates writing later patches here to simplify the logic
for conda-forge.
---
 .../cling/lib/Interpreter/CIFactory.cpp       | 61 +++++++------------
 1 file changed, 21 insertions(+), 40 deletions(-)

diff --git a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
index c0c73026..16659c29 100644
--- a/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
+++ b/src/interpreter/cling/lib/Interpreter/CIFactory.cpp
@@ -93,13 +93,11 @@ namespace {
   }
 
   class AdditionalArgList {
-    typedef std::deque< std::pair<const char*,std::string> > container_t;
+    typedef std::vector< std::pair<const char*,std::string> > container_t;
     container_t m_Saved;
 
   public:
-    void preArgument(const char* arg, std::string value = std::string()) {
-      m_Saved.push_front(std::make_pair(arg,std::move(value)));
-    }
+    
     void addArgument(const char* arg, std::string value = std::string()) {
       m_Saved.push_back(std::make_pair(arg,std::move(value)));
     }
@@ -113,23 +111,16 @@ namespace {
   static void ReadCompilerIncludePaths(const char* Compiler,
                                        llvm::SmallVectorImpl<char>& Buf,
                                        AdditionalArgList& Args,
-                                       const std::string& resourcePath,
                                        bool Verbose) {
     std::string CppInclQuery("LC_ALL=C ");
     CppInclQuery.append(Compiler);
 
     CppInclQuery.append(" -xc++ -E -v /dev/null 2>&1 |"
-                        " sed -n -e '/^.*include/,${' -e '/^ \\/.*include/p' -e '}'");
+                        " sed -n -e '/^.*include/,${' -e '/^ \\/.*++/p' -e '}'");
 
     if (Verbose)
       cling::log() << "Looking for C++ headers with:\n  " << CppInclQuery << "\n";
 
-    std::string resourceIncDir;
-    llvm::SmallString<512> recInc(resourcePath);
-    llvm::sys::path::append(recInc, "include");
-    if (llvm::sys::fs::exists(recInc.str()))
-      resourceIncDir = recInc.str().str();
-
     if (FILE *PF = ::popen(CppInclQuery.c_str(), "r")) {
       Buf.resize(Buf.capacity_in_bytes());
       while (fgets(&Buf[0], Buf.capacity_in_bytes(), PF) && Buf[0]) {
@@ -141,15 +132,8 @@ namespace {
             if (Verbose)
               cling::utils::LogNonExistantDirectory(Path);
           }
-          else {
-            if (!resourceIncDir.empty()) {
-              llvm::SmallString<512> tmp(Path);
-              llvm::sys::path::append(tmp, "xmmintrin.h");
-              if (llvm::sys::fs::exists(tmp.str()))
-                Args.addArgument("-cxx-isystem", resourceIncDir);
-            }
+          else
             Args.addArgument("-cxx-isystem", Path.str());
-          }
         }
       }
       ::pclose(PF);
@@ -227,21 +211,6 @@ namespace {
     static AdditionalArgList sArguments;
     if (sArguments.empty()) {
       const bool Verbose = opts.Verbose;
-      std::string resourcePath;
-
-      if (!opts.ResourceDir && !opts.NoBuiltinInc) {
-        resourcePath = getResourceDir(llvmdir);
-
-        // FIXME: Handle cases, where the cling is part of a library/framework.
-        // There we can't rely on the find executable logic.
-        if (!llvm::sys::fs::is_directory(resourcePath)) {
-          cling::errs()
-            << "ERROR in cling::CIFactory::createCI():\n  resource directory "
-            << resourcePath << " not found!\n";
-          resourcePath = "";
-        }
-      }
-
 #ifdef _MSC_VER
       // When built with access to the proper Windows APIs, try to actually find
       // the correct include paths first. Init for UnivSDK.empty check below.
@@ -332,14 +301,14 @@ namespace {
             clang.append(" -stdlib=libstdc++");
   #endif
           }
-          ReadCompilerIncludePaths(clang.c_str(), buffer, sArguments, resourcePath, Verbose);
+          ReadCompilerIncludePaths(clang.c_str(), buffer, sArguments, Verbose);
         }
   #endif // _LIBCPP_VERSION
 
   // First try the relative path 'g++'
   #ifdef CLING_CXX_RLTV
         if (sArguments.empty())
-          ReadCompilerIncludePaths(CLING_CXX_RLTV, buffer, sArguments, resourcePath, Verbose);
+          ReadCompilerIncludePaths(CLING_CXX_RLTV, buffer, sArguments, Verbose);
   #endif
   // Then try the include directory cling was built with
   #ifdef CLING_CXX_INCL
@@ -349,7 +318,7 @@ namespace {
   // Finally try the absolute path i.e.: '/usr/bin/g++'
   #ifdef CLING_CXX_PATH
         if (sArguments.empty())
-          ReadCompilerIncludePaths(CLING_CXX_PATH, buffer, sArguments, resourcePath, Verbose);
+          ReadCompilerIncludePaths(CLING_CXX_PATH, buffer, sArguments, Verbose);
   #endif
 
         if (sArguments.empty()) {
@@ -439,8 +408,20 @@ namespace {
 
 #endif // _MSC_VER
 
-      if (!resourcePath.empty())
-        sArguments.addArgument("-resource-dir", std::move(resourcePath));
+      if (!opts.ResourceDir && !opts.NoBuiltinInc) {
+        std::string resourcePath = getResourceDir(llvmdir);
+
+        // FIXME: Handle cases, where the cling is part of a library/framework.
+        // There we can't rely on the find executable logic.
+        if (!llvm::sys::fs::is_directory(resourcePath)) {
+          cling::errs()
+            << "ERROR in cling::CIFactory::createCI():\n  resource directory "
+            << resourcePath << " not found!\n";
+          resourcePath = "";
+        } else {
+          sArguments.addArgument("-resource-dir", std::move(resourcePath));
+        }
+      }
     }
 
     for (auto& arg : sArguments) {
-- 
2.52.0

